userspace/engine/falco_engine.cpp

```diff
@@ -335,6 +347,11 @@ void falco_engine::enable_rule_by_tag(const std::set<std::string> &tags, bool en
 {
 	uint16_t ruleset_id = find_ruleset_id(ruleset);
 
+	enable_rule_by_tag(tags, enabled, ruleset_id);
+}
+
+void falco_engine::enable_rule_by_tag(const std::set<std::string> &tags, bool enabled, const uint16_t ruleset_id)
```

Comment:
@mstemm could we add and/or expand unit tests? We recently have been more rigorous wrt always adding new tests when touching any part of the code base especially in libs.

Touching these parts of the code base reminded me of a closed PR https://github.com/falcosecurity/falco/pull/2671/files (will be reworked later by Jason). 

I wasn't sure then if enabling by tags was really fixed? What if we enable a rule by a tag, but the rule is disabled by `enabled: false` -> what takes precedence in such a case? @jasondellaluce do you remember the Falco issue around issues with enabling by tags? We might as well resolve this here in this PR especially as we expand test cases either here or in the falco testing Go repo.

Reply:
Hrm, there used to be tests related to loading rules with enabled/disabled flags, but they all were removed: https://github.com/falcosecurity/falco/commit/b1ae5d3dd9f0cc19abfc306aa4110b04d614be8e#diff-b0d96f583cb4d5e7f6ef87d9614e3826058db099ea1c6cff8a6af305c609ce4cL605. It doesn't cover this specific case of using tags, but a lot of other combinations were already tested. Did those tests get migrated to google test?

I can add unit tests for loading rules although I don't actually see any tests for that right now below unit_tests. Is there an example I can follow?

Reply:
I see that there's https://github.com/falcosecurity/testing. I guess I should add tests there?

Reply:
@mstemm up to you re the How To. If we instead need some traditional unit tests here I would appreciate them especially for the case of enabling a rule by a tag when the rule was explicitly disabled via the `enabled` key. Also aware I am asking for more tests a bit outside of the scope of this PR, hope that is ok?

Reply:
I added unit tests for enabling/disabling rules: unit_tests/engine/test_enable_rule.cpp.

Reply:
Ok added test for tags as well as the specific case of enabling a disabled rule by tag for the default ruleset.

Reply:
Thanks @mstemm! It's a start for unit testing. Few more thoughts: 

It would be nice to have more tests in this regard at the app actions level for `falco::app::run_result falco::app::actions::load_rules_files(falco::app::state& s)` for instance with some more realistic rules (that is: a few more rules that combine all the possibilities we can think of in terms of for example has `enabled: false` but also a tag that would enable the rule and also a tag that disables that rule and all sorts of combinations). That way we would clearly test what has precedence, meaning what enables or disables a rule at the end as that's what can really break things. WDYT? We could do that here (I would appreciate it) or would you prefer to defer this (create a ticket so we don't forget about it)? I am not intending to block this PR.

Reply:
I think a lot of those cases are already covered in https://github.com/falcosecurity/testing/blob/main/tests/falco/legacy_test.go, aren't they?

Reply:
I see bunch of tests that indeed sound about right, however what are we asserting? Just checking if exit code is 0 and no errors? Not seeing any assertion(s) of whether the rule(s) were enabled or disabled in the end, do you?

By the way the docs probably could be updated as well with more clarity around the use of enabling/disabling via tags vs the enabled key https://falco.org/docs/rules/controlling-rules/#tags

Reply:
Depends on the test, but they do check specific rules. For example https://github.com/falcosecurity/testing/blob/main/tests/falco/legacy_test.go#L385 has a bunch of rules with specific tags, enables some tags, and verifies the rules that should match the sample event.

Reply:
ACK, left a comment in the existing open issue https://github.com/falcosecurity/testing/issues/7#issuecomment-1832811140, worth checking in the future.

So we consider this PR done then right? Or anything you wanted to add?

Reply:
I didn't have anything else to add.